libminifi/src/Repository.cpp

```diff
@@ -0,0 +1,140 @@
+/**
+ * @file Repository.cpp
+ * Repository implemenatation 
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <cstdint>
+#include <vector>
+#include <arpa/inet.h>
+#include "io/DataStream.h"
+#include "io/Serializable.h"
+#include "Relationship.h"
+#include "Logger.h"
+#include "FlowController.h"
+#include "Repository.h"
+#include "Provenance.h"
+#include "FlowFileRepository.h"
+
+const char *Repository::RepositoryTypeStr[MAX_REPO_TYPE] = {"Provenace Repository", "FlowFile Repository"};
+uint64_t Repository::_repoSize[MAX_REPO_TYPE] = {0, 0}; 
+
+void Repository::start() {
+	if (!_enable)
+		return;
+	if (this->_purgePeriod <= 0)
+		return;
+	if (_running)
+		return;
+	_running = true;
+	logger_->log_info("%s Repository Monitor Thread Start", RepositoryTypeStr[_type]);
+	_thread = new std::thread(run, this);
+	_thread->detach();
+}
+
+void Repository::stop() {
+	if (!_running)
+		return;
+	_running = false;
+	logger_->log_info("%s Repository Monitor Thread Stop", RepositoryTypeStr[_type]);
+}
+
+void Repository::run(Repository *repo) {
+	// threshold for purge
```

Comment:
uint64_t purgeThreshold = repo->_maxPartitionBytes * 3 / 4;

# Provenance Repository #
nifi.provenance.repository.directory.default=./provenance_repository
nifi.provenance.repository.max.storage.time=1 MIN
nifi.provenance.repository.max.storage.size=1 MB
So once the repo size exceed 75%, we start to purge the record if the record was stay in the repo exceed the max storage time.
If repo size is full, we mark the repo full and we will stop to put the record in repo.